Bluetooth: Mesh: Add user guide to implement vendor model (#3159)
nrfconnect/sdk-nrf: samples/bluetooth/mesh/chat/README.rst

.. _bt_mesh_chat:

> Bluetooth: Mesh Chat

Reviewer: Mesh chat

####################

> The Bluetooth Mesh Chat sample demonstrates how the mesh network can be used to facilitate communication between nodes by text, using the :ref:`bt_mesh_chat_client_model`.

Reviewer: Bluetooth mesh chat sample

By means of the mesh network, the clients as mesh nodes can communicate with each other without the need of a server.
The sample is mainly designed for group communication, but it also supports one-on-one communication, as well as sharing the nodes presence.

> This sample is used in :ref:`ug_bt_mesh_vendor_model` as an example of how to implement a vendor model for the Bluetooth Mesh in |NCS|.

Reviewer: Bluetooth mesh

.. toctree::
   :maxdepth: 1
   :glob:
   :caption: Subpages:

   chat_cli.rst

Overview
********

This sample is split into four source files:

* A :file:`main.c` file to handle initialization.
* A file for handling the Chat Client model, :file:`chat_cli.c`.

> * A file for handling Bluetooth Mesh models and communication with the :ref:`shell module <shell_api>`, :file:`model_handler.c`.

Reviewer: Bluetooth mesh models

> After provisioning and configuring the Bluetooth Mesh models supported by the sample in the `nRF Mesh mobile app`_, you can communicate with other mesh nodes by sending text messages and obtaining their presence using the :ref:`shell module <shell_api>`.

Reviewer: Bluetooth mesh models

Provisioning
============

The provisioning is handled by the :ref:`bt_mesh_dk_prov`.

Models
======

> The following table shows the Bluetooth Mesh Chat composition data for this sample:

Reviewer: Bluetooth mesh chat composition data

   Config Server
   Health Server
   Chat Client
   =================

The models are used for the following purposes:

* :ref:`bt_mesh_chat_cli_readme` instance in the first element is used to communicate with other Chat Client models instantiated on other mesh nodes.
* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.

The model handling is implemented in :file:`src/model_handler.c`.

Requirements
************

* At least two of the following development kits:

  * |nRF5340DK|; if you use this development kit, add the following options to the configuration of the network sample:

    .. code-block:: none

        CONFIG_BT_CTLR_TX_BUFFER_SIZE=74
        CONFIG_BT_CTLR_DATA_LENGTH_MAX=74

    This is required because Bluetooth Mesh has different |BLE| Controller requirements than other Bluetooth samples.

  * |nRF52840DK|
  * |nRF52DK|

>   * |nRF51DK|

Reviewer 2: No, it's not officially supported

.. _bluetooth_mesh_chat:

Bluetooth: Mesh Chat
####################

The Bluetooth Mesh Chat sample demonstrates how the mesh network can be used to facilitate communication between nodes by text, using the :ref:`bluetooth_mesh_chat_client_model`.
By means of the mesh network, the clients as mesh nodes can communicate with each other without the need of a server.
The sample is mainly designed for group communication, but it also supports one-on-one communication, as well as sharing the nodes presence.

This sample is used in :ref:`ug_bt_mesh_vendor_model` as an example of how to implement a vendor model for the Bluetooth Mesh in |NCS|.

Overview
********

This sample is split into four source files:

* A :file:`main.c` file to handle initialization.
* A file for handling the Chat Client model, :file:`chat_cli.c`.
* A file for handling Bluetooth Mesh models, :file:`model_handler.c`.

> * A file for handling communication with UART, :file:`uart_handler.c`.

Reviewer 2: Remove uart handler

> After provisioning and configuring the Bluetooth Mesh models supported by the sample in the `nRF Mesh mobile app`_, you can communicate with other mesh nodes by sending text messages and obtaining their presence.

Reviewer 2: Add a sentence about the usage of the Shell module?

Provisioning
============

The provisioning is handled by the :ref:`bt_mesh_dk_prov`.

Models
======

The following table shows the Bluetooth Mesh Chat composition data for this sample:

   +---------------+
   |  Element 1    |
   +===============+
   | Config Server |
   +---------------+
   | Health Server |
   +---------------+
   | Chat Client   |
   +---------------+

The models are used for the following purposes:

* The :ref:`bluetooth_mesh_chat_client_model` instance in the first element is used to communicate with the other Chat Client models instantiated on the other mesh nodes.
* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.

The model handling is implemented in :file:`src/model_handler.c`.

.. _bluetooth_mesh_chat_client_model:

Chat Client model

> -----------------

Reviewer 2: I think this should be split out to a separate subpage that follows the model template. It's mostly this first paragraph and the states and messages that fits there though.

The Chat Client model is a vendor model that allows communication with other such models, by sending text messages and providing the presence of the model instance.
It demonstrates basics of a vendor model implementation.
The model doesn't have a limitation on per-node instantiations of the model, and therefore can be instantiated on each element of the node.

Adding the model to a node composition data
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The Chat Client model is a vendor model, and therefore in the application, when defining the node composition data, it needs to be declared in the third argument in :c:macro:`BT_MESH_ELEM` macro:

.. code-block:: c

    static struct bt_mesh_elem elements[] = {
        BT_MESH_ELEM(1,
            BT_MESH_MODEL_LIST(
                BT_MESH_MODEL_CFG_SRV(&cfg_srv),
                BT_MESH_MODEL_HEALTH_SRV(&health_srv, &health_pub)),
            BT_MESH_MODEL_LIST(
                BT_MESH_MODEL_CHAT_CLI(&chat))),
    };

.. _bluetooth_mesh_chat_client_model_message_sending:

Message sending
^^^^^^^^^^^^^^^

> For sending a text message, the Chat Client model implements both ways defined by :ref:`zephyr:bluetooth_mesh_access`.

Reviewer 2: ```suggestion
The Chat Client model implements both approaches to message sending using the :ref:`zephyr:bluetooth_mesh_access`.
```

   Presence message has a defined length of 1 byte.

Presence Get
   Used to retrieve the current model presence.
   Upon receiving the Presence Get message, the Chat Client model will send the Presence message with the current model presence stored in the response.
   The message doesn't have any payload.

Message
   Used to send a non-private text message.
   The payload consists of the text string terminated by ``\0``.
   The length of the text string can be configured at the compile-time using `BT_MESH_CHAT_CLI_MESSAGE_LENGTH` option.

Private Message
   Used to send a private text message.
   When the model receives this message, it replies with the Message Reply.
   The payload consists of the text string terminated by ``\0``.
   The length of the text string can be configured at the compile-time using `BT_MESH_CHAT_CLI_MESSAGE_LENGTH` option.

Message Reply
   Used to reply on the received Private Message to confirm the reception.
   The message doesn't have any payload.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading

>    :rows: nrf5340dk_nrf5340_cpuapp_and_cpuappns, nrf52840dk_nrf52840, nrf52dk_nrf52832, nrf52833dk_nrf52833, nrf52833dk_nrf52820

Reviewer 2: ```suggestion
   :sample-yaml-rows:
```

.. note::
   If you use nRF5340 PDK, add the following options to the configuration of the network sample:

   .. code-block:: none

      CONFIG_BT_CTLR_TX_BUFFER_SIZE=74
      CONFIG_BT_CTLR_DATA_LENGTH_MAX=74

>       CONFIG_BT_LL_SW_SPLIT=y

Reviewer 2: This is no longer necessary
